### PR TITLE
WarpX 23.07

### PR DESCRIPTION
**Pull request by ax3l**
Update WarpX and related Python packages to the lastest releases.

### CODE REVIEW

**Review by tldahlgren — DISMISSED** [review]
Confirmed the new version sha256, rest LGTM.

**Comment by tldahlgren** [comment]
Looks like the PR fails to build for two CI stacks due to `py-picmistandard`.

**Comment by adamjstewart** [comment]
@ax3l can you verify the checksums for picmicstandard and see where the incorrect ones are coming from? Might just be from GitHub instead of PyPI.

**Review by adamjstewart — DISMISSED** [review]
Changes look good, just need to fix checksum

**Comment by ax3l** [comment]
Checksums fixed! :tada: @tldahlgren @adamjstewart

**Review by tldahlgren — DISMISSED** [review]
Confirmed all of the affected version sha256.  Thanks!

File: var/spack/repos/builtin/packages/py-picmistandard/package.py [hunk 1]
```diff
@@ -11,27 +11,20 @@ class PyPicmistandard(PythonPackage):
...
     version("0.0.20", sha256="9c1822eaa2e4dd543b5afcfa97940516267dda3890695a6cf9c29565a41e2905")
     version("0.0.19", sha256="4b7ba1330964fbfd515e8ea2219966957c1386e0896b92d36bd9e134afb02f5a")
     version("0.0.18", sha256="68c208c0c54b4786e133bb13eef0dd4824998da4906285987ddee84e6d195e71")
-    # 0.15 - 0.17 have broken install logic: missing requirements.txt on pypi
-    version(
-        "0.0.16",
-        sha256="b7eefdae1c43119984226b2df358c86fdeef7495084e47b3575e3d07e790ba30",
-        url="https://github.com/picmi-standard/picmi/archive/refs/tags/0.0.14.tar.gz",
-    )
-    version("0.0.14", sha256="8f83b25b281fc0309a0c4f75c7605afd5fa0ef4df3b3ac115069478c119bc8c3")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-picmistandard/package.py:34]: Technically we should deprecate these before removing them

Comment (ax3l) [reply var/spack/repos/builtin/packages/py-picmistandard/package.py:34]: Technically yes. But I think it's ok here - we really neither use nor support the first versions that had these install bugs and no proper tags.

Also, I do not have time to figure out the Spack fetcher bug, that decides to pick random git tags for some of the releases the moment I have a single version in here that was specified by git.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-picmistandard/package.py:34]: You're the maintainer, so the final decision is up to you 😄

File: var/spack/repos/builtin/packages/py-picmistandard/package.py [hunk 2]
```diff
@@ -11,27 +11,20 @@ class PyPicmistandard(PythonPackage):
 
     homepage = "https://picmi-standard.github.io"
     git = "https://github.com/picmi-standard/picmi.git"
-    pypi = "picmistandard/picmistandard-0.0.18.tar.gz"
+    pypi = "picmistandard/picmistandard-0.25.0.tar.gz"
 
     maintainers("ax3l", "dpgrote", "RemiLehe")
 
     version("develop", branch="master")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-picmistandard/package.py:18]:
```suggestion
    version("master", branch="master")
```

File: var/spack/repos/builtin/packages/py-picmistandard/package.py [hunk 3]
```diff
@@ -11,27 +11,20 @@ class PyPicmistandard(PythonPackage):
...
     version("0.0.18", sha256="68c208c0c54b4786e133bb13eef0dd4824998da4906285987ddee84e6d195e71")
-    # 0.15 - 0.17 have broken install logic: missing requirements.txt on pypi
-    version(
-        "0.0.16",
-        sha256="b7eefdae1c43119984226b2df358c86fdeef7495084e47b3575e3d07e790ba30",
-        url="https://github.com/picmi-standard/picmi/archive/refs/tags/0.0.14.tar.gz",
-    )
-    version("0.0.14", sha256="8f83b25b281fc0309a0c4f75c7605afd5fa0ef4df3b3ac115069478c119bc8c3")
 
     depends_on("python@3.6:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-picmistandard/package.py:29]: You could remove this, Spack only supports 3.7+ anyway

File: var/spack/repos/builtin/packages/py-warpx/package.py [hunk 4]
```diff
@@ -94,14 +78,13 @@ class PyWarpx(PythonPackage):
...
     depends_on("py-picmistandard@0.0.19", type=("build", "run"), when="@22.02:22.09")
     depends_on("py-picmistandard@0.0.20", type=("build", "run"), when="@22.10:22.11")
     depends_on("py-picmistandard@0.0.22", type=("build", "run"), when="@22.12:23.03")
     depends_on("py-picmistandard@0.23.2", type=("build", "run"), when="@23.04:23.05")
-    depends_on("py-picmistandard@0.24.0", type=("build", "run"), when="@23.06:")
+    depends_on("py-picmistandard@0.24.0", type=("build", "run"), when="@23.06")
+    depends_on("py-picmistandard@0.25.0", type=("build", "run"), when="@23.07:")
     depends_on("py-setuptools@42:", type="build")
     # Since we use PYWARPX_LIB_DIR to pull binaries out of the
     # 'warpx' spack package, we don't need py-cmake as declared
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-warpx/package.py:90]: Wheel is already added by the base class and could be removed here

File: var/spack/repos/builtin/packages/py-warpx/package.py [hunk 5]
```diff
@@ -94,14 +78,13 @@ class PyWarpx(PythonPackage):
     depends_on("py-numpy@1.15.0:1", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-warpx/package.py:78]: Python 3.7+ is the minimum version we support, so 3.6+ could be removed

File: var/spack/repos/builtin/packages/py-warpx/package.py [hunk 6]
```diff
@@ -94,14 +78,13 @@ class PyWarpx(PythonPackage):
...
-    depends_on("py-picmistandard@0.0.14", type=("build", "run"), when="@21.03:21.11")
-    depends_on("py-picmistandard@0.0.16", type=("build", "run"), when="@21.12")
     depends_on("py-picmistandard@0.0.18", type=("build", "run"), when="@22.01")
     depends_on("py-picmistandard@0.0.19", type=("build", "run"), when="@22.02:22.09")
     depends_on("py-picmistandard@0.0.20", type=("build", "run"), when="@22.10:22.11")
     depends_on("py-picmistandard@0.0.22", type=("build", "run"), when="@22.12:23.03")
     depends_on("py-picmistandard@0.23.2", type=("build", "run"), when="@23.04:23.05")
-    depends_on("py-picmistandard@0.24.0", type=("build", "run"), when="@23.06:")
+    depends_on("py-picmistandard@0.24.0", type=("build", "run"), when="@23.06")
+    depends_on("py-picmistandard@0.25.0", type=("build", "run"), when="@23.07:")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-warpx/package.py:87]: I try to sort these from newest to oldest to match the version directive

**Review by adamjstewart — APPROVED** [review]
Some optional comments, feel free to merge if you think it's good enough

File: var/spack/repos/builtin/packages/py-picmistandard/package.py (hunk 3 above)
Comment (ax3l) [reply var/spack/repos/builtin/packages/py-picmistandard/package.py:29]: Ok. Next release is 3.8+

**Comment by ax3l** [comment]
Thanks! I will merge this and add the suggestions to the 23.08 release #39407 :)

**Pull request merged**